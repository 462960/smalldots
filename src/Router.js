import { PureComponent, PropTypes } from 'react'
import createHistory from 'history/createBrowserHistory'
import qs from 'qs'
import isPlainObject from 'lodash/isPlainObject'
import find from 'lodash/find'
import Route from 'route-parser'

let history = null
if (typeof document !== 'undefined') {
  history = createHistory()
}

class Router extends PureComponent {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.func
    ]).isRequired
  }

  unlisten = history.listen(location => {
    if (this.willUnmount) {
      return
    }
    this.forceUpdate()
  })

  componentWillUnmount() {
    this.willUnmount = true
    this.unlisten()
  }

  push = (path, state) => history.push(path, state)

  replace = (path, state) => history.replace(path, state)

  go = n => history.go(n)

  back = () => history.goBack()

  forward = () => history.goForward()

  render() {
    const props = {
      pathname: history.location.pathname,
      search: qs.parse(history.location.search.substr(1)),
      hash: (
        history.location.hash.match('=')
          ? qs.parse(history.location.hash)
          : history.location.hash.replace('#', '')
      ),
      state: history.location.state || {},
      entries: history.entries,
      push: this.push,
      replace: this.replace,
      go: this.go,
      back: this.back,
      forward: this.forward
    }
    if (isPlainObject(this.props.children)) {
      const routes = Object.keys(this.props.children)
      const match = find(routes, route => (
        new Route(route).match(history.location.pathname)
      ))
      if (!match) {
        return null
      }
      return this.props.children[match]({
        ...props,
        params: new Route(match).match(history.location.pathname)
      })
    }
    return this.props.children(props) || null
  }
}

export default Router
