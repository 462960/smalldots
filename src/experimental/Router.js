import { Component, PropTypes, isValidElement } from 'react'
import createHistory from 'history/createBrowserHistory'
import Route from 'route-parser'
import queryString from 'query-string'
import isPlainObject from 'lodash/isPlainObject'

export default class Router extends Component {
  static propTypes = {
    basename: PropTypes.string,
    forceRefresh: PropTypes.bool,
    children: PropTypes.func.isRequired
  }

  state = { currentLocation: null }

  history = createHistory({
    basename: this.props.basename,
    forceRefresh: this.props.forceRefresh
  })

  unlisten = this.history.listen(location => {
    if (this.willUnmount) {
      return
    }
    this.setState({ currentLocation: location })
  })

  componentWillMount() {
    this.setState({ currentLocation: this.history.location })
  }

  componentWillUnmount() {
    this.willUnmount = true
    this.unlisten()
  }

  getHistory = () => this.history.entries

  push = (path, state) => this.history.push(path, state)

  replace = (path, state) => this.history.replace(path, state)

  go = n => this.history.go(n)

  back = () => this.history.goBack()

  forward = () => this.history.goForward()

  render() {
    const children = this.props.children({
      currentLocation: this.state.currentLocation,
      getHistory: this.getHistory,
      push: this.push,
      replace: this.replace,
      go: this.go,
      back: this.back,
      forward: this.forward
    })
    if (!isPlainObject(children)) {
      throw new Error('children should return a plain object')
    }
    const path = this.state.currentLocation.pathname
    const match = Object.keys(children).find(path => (
      new Route(path).match(path)
    ))
    if (!match) {
      return null
    }
    if (typeof children[match] !== 'function') {
      throw new Error(`${match} should be a function that returns an element`)
    }
    const element = children[match]({
      path,
      params: new Route(match).match(path),
      search: queryString.parse(this.state.currentLocation.search),
      hash: (
        this.state.currentLocation.hash.match('=')
          ? queryString.parse(this.state.currentLocation.hash)
          : this.state.currentLocation.hash.replace('#', '')
      ),
      state: this.state.currentLocation.state || {}
    })
    if (!isValidElement(element)) {
      throw new Error(`${match} should return a valid element`)
    }
    return element
  }
}
