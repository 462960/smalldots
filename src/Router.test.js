import React from 'react'
import renderer from 'react-test-renderer'
import Router from './Router'

test('renders correctly when children is a string', () => {
  const tree = renderer.create(
    <Router>
      {() => null}
    </Router>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders correctly when children is an object', () => {
  const tree = renderer.create(
    <Router>
      {{
        '/': () => null
      }}
    </Router>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
