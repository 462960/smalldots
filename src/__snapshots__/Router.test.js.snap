exports[`test renders correctly when children is a string 1`] = `null`;

exports[`test renders correctly when children is an object 1`] = `null`;
